src/timer.rs

```diff
@@ -225,3 +225,104 @@ pub trait GeneralPurposeTimer {
+            $new:ident,
+            $apbenr:ident, $apbrstr:ident,
+            $master_en:ident, $slave_en:ident,
+            $master_rst:ident, $slave_rst:ident,
+            $mms:ty, $sms:ty, $ts:expr
+        ),
+    )+) => {
+        $(
+            impl LinkedTimerPair<$MASTER, $SLAVE> {
+                /// Create and configure a new `LinkedTimerPair` with the
+                /// specified timers.
+                pub fn $new(tim_master: $MASTER, tim_slave: $SLAVE, rcc: &mut Rcc) -> Self {
+                    // Enable timers
+                    rcc.rb.$apbenr.modify(|_, w| w.$master_en().set_bit());
+                    rcc.rb.$apbenr.modify(|_, w| w.$slave_en().set_bit());
+
+                    // Reset timers
+                    rcc.rb.$apbrstr.modify(|_, w| w.$master_rst().set_bit());
+                    rcc.rb.$apbrstr.modify(|_, w| w.$master_rst().clear_bit());
+                    rcc.rb.$apbrstr.modify(|_, w| w.$slave_rst().set_bit());
+                    rcc.rb.$apbrstr.modify(|_, w| w.$slave_rst().clear_bit());
+
+                    // Enable counter
+                    tim_master.cr1.modify(|_, w| w.cen().set_bit());
+                    tim_slave.cr1.modify(|_, w| w.cen().set_bit());
+
+                    // In the MMS (Master Mode Selection) register, set the master mode so
+                    // that a rising edge is output on the trigger output TRGO every time
+                    // an update event is generated.
+                    tim_master.cr2.modify(|_, w| w.mms().variant(<$mms>::UPDATE));
+
+                    // In the SMCR (Slave Mode Control Register), select the
+                    // appropriate internal trigger source (TS).
+                    tim_slave.smcr.modify(|_, w| w.ts().variant($ts));
+                    // Set the SMS (Slave Mode Selection) register to "external clock mode 1",
+                    // where the rising edges of the selected trigger (TRGI) clock the counter.
+                    tim_slave.smcr.modify(|_, w| w.sms().variant(<$sms>::EXT_CLOCK_MODE));
+
+                    Self { tim_master, tim_slave }
+                }
+            }
+
+            impl LinkedTimer for LinkedTimerPair<$MASTER, $SLAVE> {
+                /// Return the current 16 bit counter value of the slave timer (MSB).
+                fn get_counter_msb(&self) -> u16 {
+                    self.tim_slave.cnt.read().cnt().bits()
+                }
+
+                /// Return the current 16 bit counter value of the master timer (LSB).
+                fn get_counter_lsb(&self) -> u16 {
+                    self.tim_master.cnt.read().cnt().bits()
+                }
+
+                /// Return the current 32 bit counter value.
+                fn get_counter(&self) -> u32 {
+                    let lsb = self.tim_master.cnt.read().cnt().bits() as u32;
+                    let msb = self.tim_slave.cnt.read().cnt().bits() as u32;
+                    (msb << 16) | lsb
```

Comment:
Assuming the timer runs at a high frequency:
1. Read master: master_cnt = 0xFFFF, slave_cnt = 0x003 (but we have not read the slave yet)
2. Read slave: master_cnt = 0x0000,  slave_cnt = 0x004
3. Return 0x0004FFFF even though the actual value should be 0x0003FFF
Not sure what’s the best way to prevent this. Maybe read the slave before and after the master counter and check if it changed

Reply:
Hmm, good point... I assume there's no way to read two registers simultaneously.

Your approach with a change check seems sensible. But what do we do if the slave changed? Retry the whole process? Return an error? Could it be a problem if reading the timer is not constant time?

Reply:
I managed to provoke the error quite easily by setting the ARR register in the LSB register to a low value like 0xff.

```
MSB mismatch: 26773 != 26774
```

I pushed a new commit that should fix the race condition.